src/xenia/base/filesystem.h

```diff
@@ -18,49 +19,53 @@
 #include "xenia/base/string.h"
 
 namespace xe {
+
+std::string path_to_utf8(const std::filesystem::path& path);
+std::u16string path_to_utf16(const std::filesystem::path& path);
+std::filesystem::path to_path(const std::string_view source);
+std::filesystem::path to_path(const std::u16string_view source);
+
 namespace filesystem {
 
 // Get executable path.
-std::wstring GetExecutablePath();
+std::filesystem::path GetExecutablePath();
 
 // Get executable folder.
-std::wstring GetExecutableFolder();
+std::filesystem::path GetExecutableFolder();
 
 // Get user folder.
-std::wstring GetUserFolder();
-
-// Canonicalizes a path, removing ..'s.
-std::string CanonicalizePath(const std::string& original_path);
+std::filesystem::path GetUserFolder();
 
 // Returns true of the specified path exists as either a directory or file.
-bool PathExists(const std::wstring& path);
+bool PathExists(const std::filesystem::path& path);
 
 // Creates the parent folder of the specified path if needed.
 // This can be used to ensure the destination path for a new file exists before
 // attempting to create it.
-bool CreateParentFolder(const std::wstring& path);
+bool CreateParentFolder(const std::filesystem::path& path);
 
 // Creates a folder at the specified path.
 // Returns true if the path was created.
-bool CreateFolder(const std::wstring& path);
+bool CreateFolder(const std::filesystem::path& path);
 
 // Recursively deletes the files and folders at the specified path.
 // Returns true if the path was found and removed.
-bool DeleteFolder(const std::wstring& path);
+bool DeleteFolder(const std::filesystem::path& path);
 
 // Returns true if the given path exists and is a folder.
-bool IsFolder(const std::wstring& path);
+bool IsFolder(const std::filesystem::path& path);
 
 // Creates an empty file at the given path.
-bool CreateFile(const std::wstring& path);
+bool CreateFile(const std::filesystem::path& path);
 
 // Opens the file at the given path with the specified mode.
 // This behaves like fopen and the returned handle can be used with stdio.
-FILE* OpenFile(const std::wstring& path, const char* mode);
+FILE* OpenFile(const std::filesystem::path& path,
+               const std::filesystem::path& mode);
```

Comment:
It's not a `path` but it's unfortunately whatever the native type is on the platform, so I cheaped out and used `path` for it. Better ideas?

Reply:
I wasn't aware there was a convenient `std::filesystem::path::string_type`. I'll do that. Thanks!

Reply:
mh, but then its not possible to call `OpenFile(path, "rb")` on Windows since `wstring` does not have a `char*` constructor like `path` does. So without braking all uses of `OpenFile()` either your hack or `std::string_view` with utf8 to utf16 conversion for windows I guess